doc/ite.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.29.
.TH SBSAT "1" "August 2003" "sbsat 2.0r-1" "User Commands"
.SH NAME
sbsat \- manual page for sbsat 2.0r-1
.SH SYNOPSIS
.B ite
[\fIOPTIONS\fR]... [\fIinputfile\fR] [\fIoutputfile\fR]
.SH DESCRIPTION
SBSat is a SAT solver.
.SH OPTIONS
.TP
\fB\-\-help\fR, \fB\-h\fR
Show all program options
.TP
\fB\-\-version\fR
Show program version
.TP
\fB\-\-create\-ini\fR
Create ini file
.TP
\fB\-\-ini\fR <string>
Set the ini file [default="~/ite.ini"]
.TP
\fB\-\-debug\fR <number>
debugging level (0-none, 9-max) [default=2]
.TP
\fB\-\-debug\-dev\fR <string>
debugging device [default="stderr"]
.TP
\fB\-\-params\-dump\fR, \fB\-D\fR
dump all internal parameters before processing
.TP
\fB\-\-input\-file\fR <string>
input filename [default="-"]
.TP
\fB\-\-output\-file\fR <string>
output filename [default="-"]
.TP
\fB\-\-temp\-dir\fR <string>
directory for temporary files [default="$TEMP"]
.TP
\fB\-\-show\-result\fR <string>, \fB\-R\fR <string>
Show result (n=no result, r=raw, f=fancy)
[default="n"]
.TP
\fB\-\-verify\-solution\fR <number>
Verify solution [default=1]
.TP
\fB\-\-expected\-result\fR <string>
Report error if the result is not as specified.
Options are SAT, UNSAT, TRIV_SAT, TRIV_UNSAT,
SOLV_SAT and SOLV_UNSAT
[default=""]
.TP
\fB\-\-comment\fR <string>
Comment to appear next to the filename  [default=""]
.TP
\fB\-\-ctrl\-c\fR <number>
Enable/Disable Ctrl-c handler to end
preproc/brancher
[default=0]
.TP
\fB\-\-reports\fR <number>
Reporting style during branching (0 - standard, 1 -
crtwin)
[default=0]
.SS "BDD options:"
.TP
\fB\-\-num\-buckets\fR <number>
Set the number of buckets in power of 2 [default=16]
.HP
\fB\-\-size\-buckets\fR <number> Set the size of a bucket in power of 2 [default=5]
.TP
\fB\-\-bdd\-pool\-size\fR <number>
The size of the bdd pool increment [default=1000000]
.SS "Input options:"
.TP
\fB\-\-limit\-and\-equ\fR <number>
The minimum # of literals to flag sp. function
and_equ
[default=2]
.TP
\fB\-\-limit\-or\-equ\fR <number> The minimum # of literals to flag sp. function
or_equ
[default=2]
.TP
\fB\-\-limit\-or\fR <number>
The minimum # of literals to flag sp. function
plainor
[default=8]
.TP
\fB\-\-limit\-xor\fR <number>
The minimum # of literals to flag sp. function
plainxor
[default=5]
.TP
\fB\-\-break\-xors\fR <number>
Break XORS into linear and non-linear functions
[default=1]
.SS "Output options:"
.TP
\fB\-b\fR
Start regular brancher [default]
.TP
\fB\-w\fR
Start walksat brancher
.TP
\fB\-m\fR
Start WVF brancher
.TP
\fB\-n\fR
Don't start any brancher or conversion
.TP
\fB\-s\fR
Output in SMURF format
.TP
\fB\-c\fR
Output in CNF format
.TP
\fB\-p\fR
Output in tree like format
.TP
\fB\-\-cnf\fR <string>
Format of CNF output (3sat, qm, noqm)
[default="noqm"]
.TP
\fB\-\-tree\fR, \fB\-tree\fR
Print tree representation
.TP
\fB\-\-tree\-width\fR <number>, \fB\-W\fR <number>
Set tree width [default=64]
.SS "Trace format options:"
.TP
\fB\-\-module\-dir\fR <string>
directory to find extra modules [default="./Modules"]
.SS "Preprocessing options:"
.TP
\fB\-\-preprocess\-sequence\fR <string>, \fB\-P\fR <string>
The preprocessing sequence
[default="(ExDc)*(ExSt)*(ExPr)*"]
.TP
\fB\-\-All\fR, \fB\-All\fR
Enable/Disable All Preprocessing Options (1/0)
.TP
\fB\-\-Cl\fR <number>, \fB\-Cl\fR <number>
Enable/Disable Clustering (1/0) [default=1]
.TP
\fB\-\-Co\fR <number>, \fB\-Co\fR <number>
Enable/Disable Cofactoring (1/0) [default=1]
.TP
\fB\-\-Pr\fR <number>, \fB\-Pr\fR <number>
Enable/Disable Pruning (1/0) [default=1]
.TP
\fB\-\-St\fR <number>, \fB\-St\fR <number>
Enable/Disable Strengthening (1/0) [default=1]
.TP
\fB\-\-In\fR <number>, \fB\-In\fR <number>
Enable/Disable Inferences (1/0) [default=1]
.TP
\fB\-\-Ex\fR <number>, \fB\-Ex\fR <number>
Enable/Disable Existential Quantification (1/0)
[default=1]
.TP
\fB\-\-Dc\fR <number>, \fB\-Dc\fR <number>
Enable/Disable Dependent Variable Clustering (1/0)
[default=1]
.TP
\fB\-\-max\-preproc\-time\fR <number>
set the time limit in seconds (0=no limit)
[default=0]
.SS "Brancher options:"
.TP
\fB\-\-lemma\-out\-file\fR <string>
File to dump lemmas to [default=""]
.TP
\fB\-\-lemma\-in\-file\fR <string>
File to read lemmas from [default=""]
.TP
\fB\-\-csv\-trace\-file\fR <string>
File to save execution trace in CSV format
[default=""]
.TP
\fB\-\-var\-stat\-file\fR <string>
File to save var stats [default=""]
.TP
\fB\-\-brancher\-presets\fR <string>
Variables that are preset before the brancher is
called. Options are ([[=|!|#|+var|-var] ]*)
[default=""]
.TP
\fB\-\-reverse\-depend\fR, \fB\-r\fR
Reverse dependency info on in/dependent variables
.TP
\fB\-\-clear\-depend\fR, \fB\-e\fR
Clear dependency information on variables
.TP
\fB\-\-heuristic\fR <string>, \fB\-H\fR <string>
Choose heuristic j=Johnson, l=Chaff-like lemma
heuristic and i=Interactive
[default="j"]
.TP
\fB\-\-backjumping\fR <number>
Enable/Disable backjumping (1/0) [default=1]
.TP
\fB\-\-max\-cached\-lemmas\fR <number>, \fB\-L\fR <number>
set the maximum # of lemmas [default=5000]
.TP
\fB\-\-max\-solutions\fR <number>
Set the maximum number of solutions to search for.
[default=1]
.TP
\fB\-\-sbj\fR <number>
Super backjumping. [default=0]
.TP
\fB\-\-max\-vbles\-per\-smurf\fR <number>, \fB\-S\fR <number>
set the maximum number variables per smurf
[default=8]
.TP
\fB\-\-backtracks\-per\-report\fR <number>
set the number of backtracks per report
[default=10000]
.TP
\fB\-\-max\-brancher\-time\fR <number>
set the time limit in seconds (0=no limit)
[default=0]
.TP
\fB\-\-max\-brancher\-cp\fR <number>
set the choice point limit (0=no limit) [default=0]
.TP
\fB\-\-brancher\-trace\-start\fR <number>
number of backtracks to start the trace (when
debug=9)
[default=0]
.TP
\fB\-\-compress\-smurfs\fR <number>
Share states among smurfs [default=1]
.TP
\fB\-\-smurfs\-share\-paths\fR <number>
Share paths among smurfs [default=0]
.SS "Johnson heuristic options:"
.TP
\fB\-\-jheuristic\-k\fR <number>, \fB\-K\fR <number>
set the value of K [default=3.000000]
.TP
\fB\-\-jheuristic\-k\-true\fR <number>
set the value of True state [default=0.000000]
.TP
\fB\-\-jheuristic\-k\-inf\fR <number>
set the value of the inference multiplier
[default=1.000000]
.SH AUTHOR
Written by a reserach team lead by John Franco.
.SH "REPORTING BUGS"
Report bugs to <mkouril@ececs.uc.edu> or <franco@gauss.ececs.uc.edu>.
.SH COPYRIGHT
Copyright \(co 1999-2003, University of Cincinnati.  All rights reserved.
.SH "SEE ALSO"
The full documentation for
.B sbsat
is maintained as a Texinfo manual.  If the
.B info
and
.B sbsat
programs are properly installed at your site, the command
.IP
.B info sbsat
.PP
should give you access to the complete manual.


ChangeLog
2010-05-05 14:40  fett

	* ChangeLog, Makefile.in, aclocal.m4, configure, doc/Makefile.in,
	  doc/man1/Makefile.in, examples/Makefile.in, include/Makefile.in,
	  include/config.h, src/Makefile.in, src/cudd/Makefile.in,
	  src/cudd/cppobj/Makefile.in, src/cudd/cudd/Makefile.in,
	  src/cudd/epd/Makefile.in, src/cudd/include/Makefile.in,
	  src/cudd/mtr/Makefile.in, src/cudd/st/Makefile.in,
	  src/cudd/util/Makefile.in, src/formats/Makefile.in,
	  src/formats/iteformat.cc, src/formats/markbdd.cc,
	  src/formats/5/Makefile.in, src/postproc/Makefile.in,
	  src/preproc/Makefile.in, src/solv_bddwalk/Makefile.in,
	  src/solv_dac/Makefile.in, src/solv_simple/Makefile.in,
	  src/solv_simple/bb_gelim/Makefile.in,
	  src/solv_simple/bb_lemmas/Makefile.in,
	  src/solv_simple/fn_inference/Makefile.in,
	  src/solv_simple/fn_minmax/Makefile.in,
	  src/solv_simple/fn_negminmax/Makefile.in,
	  src/solv_simple/fn_or/Makefile.in,
	  src/solv_simple/fn_smurf/Makefile.in,
	  src/solv_simple/fn_xor/Makefile.in,
	  src/solv_simple/include/Makefile.in, src/solv_smurf/Makefile.in,
	  src/solv_smurf/fn_and/Makefile.in,
	  src/solv_smurf/fn_lemma/Makefile.in,
	  src/solv_smurf/fn_lemma_au/Makefile.in,
	  src/solv_smurf/fn_minmax/Makefile.in,
	  src/solv_smurf/fn_smurf/Makefile.in,
	  src/solv_smurf/fn_smurf_au/Makefile.in,
	  src/solv_smurf/fn_smurf_xor/Makefile.in,
	  src/solv_smurf/fn_xor/Makefile.in,
	  src/solv_smurf/fn_xor_smurf/Makefile.in,
	  src/solv_smurf/hr_anne/Makefile.in,
	  src/solv_smurf/hr_berkmin/Makefile.in,
	  src/solv_smurf/hr_lemma/Makefile.in,
	  src/solv_smurf/hr_lsgb/Makefile.in,
	  src/solv_smurf/hr_lsgb_lemma/Makefile.in,
	  src/solv_smurf/hr_lsgbw/Makefile.in,
	  src/solv_smurf/hr_minisat/Makefile.in,
	  src/solv_smurf/hr_null/Makefile.in,
	  src/solv_smurf/hr_vsids/Makefile.in,
	  src/solv_smurf/include/Makefile.in, src/solv_wvf/Makefile.in,
	  src/utils/Makefile.in, tests/Makefile.in, tests/cnf/Makefile.in,
	  tests/longer_tests/Makefile.in, tests/trace_tests/Makefile.in:
	  [no log message]

2010-04-20 08:36  jeremy

	* src/generator/gentest.cc: fix char* warnings

2010-04-19 11:00  jeremy

	* src/generator/slider3_base.py: negate state3 to produce
	  alternating states in slider3

2010-04-18 10:04  fett

	* include/config.h, src/solv_simple/fn_inference/fn_inference.cc,
	  src/solv_simple/fn_minmax/fn_minmax.cc,
	  src/solv_simple/fn_negminmax/fn_negminmax.cc,
	  src/solv_simple/include/smurfs.h: Finished .dot printing for
	  cardinality constraints.

	     ~Sean

2010-04-12 21:16  jeremy

	* src/formats/: bdd2cnf.cc, sat.cc, smurf.cc: fix warnings

2010-04-12 21:09  jeremy

	* src/cudd/util/test-res.c: fix warnings

2010-04-12 00:15  fett

	* include/config.h, src/formats/markbdd.cc,
	  src/formats/5/libt5_a-aig_g.h: [no log message]

2010-04-12 00:00  fett

	* include/queue.h, src/solv_simple/fn_minmax/fn_minmax.cc,
	  src/solv_simple/fn_negminmax/fn_negminmax.cc,
	  src/solv_simple/fn_smurf/fn_smurf.cc,
	  src/solv_simple/include/fn_minmax.h,
	  src/solv_simple/include/fn_negminmax.h, src/utils/queue.cc: [no
	  log message]

2010-04-11 16:41  fett

	* src/solv_simple/include/solver.h: [no log message]

2010-04-06 11:24  jeremy

	* include/sbsat.h: Change include (deprecated warning)

2010-04-06 09:38  jeremy

	* src/generator/vanDerWaerden.cc: fix string to char* warnings

2010-04-04 18:06  fett

	* src/formats/: iteformat.cc, markbdd.cc: Added an error message
	  for '!' as an invalid character in the markbdd format.

2010-03-27 19:33  fett

	* src/formats/cnf.cc: [no log message]

2010-02-08 15:31  fett

	* include/sbsat_solver.h, src/solv_simple/solv_simple.cc,
	  src/solv_simple/solv_simple_init.cc,
	  src/solv_simple/include/solver.h, src/utils/params.cc: [no log
	  message]

2010-01-22 09:49  fett

	* Makefile.in, aclocal.m4, configure, doc/Makefile.in,
	  doc/man1/Makefile.in, examples/Makefile.in, include/Makefile.am,
	  include/Makefile.in, include/config.h, include/config.h.in,
	  include/sbsat.h, include/sbsat_vars.h, src/Makefile.in,
	  src/cudd/Makefile.in, src/cudd/cppobj/Makefile.in,
	  src/cudd/cudd/Makefile.in, src/cudd/epd/Makefile.in,
	  src/cudd/include/Makefile.in, src/cudd/mtr/Makefile.in,
	  src/cudd/st/Makefile.in, src/cudd/util/Makefile.in,
	  src/formats/Makefile.in, src/formats/5/Makefile.in,
	  src/postproc/Makefile.in, src/preproc/Makefile.in,
	  src/solv_bddwalk/Makefile.in, src/solv_dac/Makefile.in,
	  src/solv_simple/Makefile.in, src/solv_simple/solv_simple.cc,
	  src/solv_simple/solv_simple_init.cc,
	  src/solv_simple/bb_gelim/Makefile.in,
	  src/solv_simple/bb_lemmas/Makefile.in,
	  src/solv_simple/fn_inference/Makefile.in,
	  src/solv_simple/fn_minmax/Makefile.in,
	  src/solv_simple/fn_negminmax/Makefile.in,
	  src/solv_simple/fn_or/Makefile.in,
	  src/solv_simple/fn_smurf/Makefile.in,
	  src/solv_simple/fn_smurf/bt_smurf.cc,
	  src/solv_simple/fn_xor/Makefile.in,
	  src/solv_simple/include/Makefile.in,
	  src/solv_simple/include/fn_smurf.h,
	  src/solv_simple/include/solver.h, src/solv_smurf/Makefile.in,
	  src/solv_smurf/fn_and/Makefile.in,
	  src/solv_smurf/fn_lemma/Makefile.in,
	  src/solv_smurf/fn_lemma_au/Makefile.in,
	  src/solv_smurf/fn_minmax/Makefile.in,
	  src/solv_smurf/fn_smurf/Makefile.in,
	  src/solv_smurf/fn_smurf_au/Makefile.in,
	  src/solv_smurf/fn_smurf_xor/Makefile.in,
	  src/solv_smurf/fn_xor/Makefile.in, src/generator/Makefile.in,
	  src/solv_smurf/fn_xor_smurf/Makefile.in,
	  src/solv_smurf/hr_anne/Makefile.in,
	  src/solv_smurf/hr_berkmin/Makefile.in,
	  src/solv_smurf/hr_lemma/Makefile.in,
	  src/solv_smurf/hr_lsgb/Makefile.in,
	  src/solv_smurf/hr_lsgb_lemma/Makefile.in,
	  src/solv_smurf/hr_lsgbw/Makefile.in,
	  src/solv_smurf/hr_minisat/Makefile.in,
	  src/solv_smurf/hr_null/Makefile.in,
	  src/solv_smurf/hr_vsids/Makefile.in,
	  src/solv_smurf/include/Makefile.in, src/solv_wvf/Makefile.in,
	  src/utils/Makefile.am, src/utils/Makefile.in,
	  src/utils/params.cc, src/utils/vars.cc, tests/Makefile.in,

src/ite.cc
/* =========FOR INTERNAL USE ONLY. NO DISTRIBUTION PLEASE ========== */ 
/*********************************************************************
 Copyright 1999-2003, University of Cincinnati.  All rights reserved.
 By using this software the USER indicates that he or she has read,
 understood and will comply with the following:
 * 
 --- University of Cincinnati hereby grants USER nonexclusive permission
 to use, copy and/or modify this software for internal, noncommercial,
 research purposes only. Any distribution, including commercial sale
 or license, of this software, copies of the software, its associated
 documentation and/or modifications of either is strictly prohibited
 without the prior consent of University of Cincinnati.  Title to copyright
 to this software and its associated documentation shall at all times
 remain with University of Cincinnati.  Appropriate copyright notice shall
 be placed on all software copies, and a complete copy of this notice
 shall be included in all copies of the associated documentation.
 No right is  granted to use in advertising, publicity or otherwise
 any trademark,  service mark, or the name of University of Cincinnati.
 * 
 * 
 --- This software and any associated documentation is provided "as is"
 * 
 UNIVERSITY OF CINCINNATI MAKES NO REPRESENTATIONS OR WARRANTIES, EXPRESS
 OR IMPLIED, INCLUDING THOSE OF MERCHANTABILITY OR FITNESS FOR A
 PARTICULAR PURPOSE, OR THAT  USE OF THE SOFTWARE, MODIFICATIONS, OR
 ASSOCIATED DOCUMENTATION WILL NOT INFRINGE ANY PATENTS, COPYRIGHTS,
 TRADEMARKS OR OTHER INTELLECTUAL PROPERTY RIGHTS OF A THIRD PARTY.
 * 
 University of Cincinnati shall not be liable under any circumstances for
 any direct, indirect, special, incidental, or consequential damages
 with respect to any claim by USER or any third party on account of
 or arising from the use, or inability to use, this software or its
 associated documentation, even if University of Cincinnati has been advised
 of the possibility of those damages.
 *********************************************************************/
/*
 *
 * SBSAT - University of Cincinnati
 *
 */
/**********************************************************
 **  main.c (S. Weaver)
 **********************************************************/  

#include "ite.h"

extern char tracer_tmp_filename[256];
extern BDDNodeStruct **original_functions;

void bdd_bdd_free_pools();
void bdd_circuit_free();
void wvfSolve();
void Verify_NoSolver(Tracer *tracer);
void Verify_Solver(Tracer *tracer);

int ite_pre_init();
void ite_main_free(Tracer *tracer);
int ite_io_init();
void ite_io_free();
int ite_preprocessing();

int check_expected_result(int result);
int ite_main_init(int argc, char *argv[]);
int ite_main_load(Tracer *&tracer);
int ite_main(Tracer *tracer);
int ite_main_preproc();

int 
main(int argc, char *argv[])
{
   int ret = NO_ERROR;
   Tracer * tracer = NULL;
   ret = ite_main_init(argc, argv);
   if (ret == NO_ERROR) {
      ret = ite_main_load(tracer);
      if (ret == NO_ERROR) {
         ret = ite_main_preproc();
         if (ret == NO_ERROR) ret = ite_main(tracer);
      }
   }
   return ite_final(ret, tracer);
}

int
ite_main_preproc()
{
   int ret = ite_preprocessing();
   if (ret == TRIV_SAT || ret == TRIV_UNSAT) return ret;
   return NO_ERROR;
}

int 
ite_main_init(int argc, char *argv[])
{
   int ret = NO_ERROR;
	
   ret = ite_pre_init();
   if (ret != NO_ERROR) return ret;
	
   ret = params_parse_cmd_line(argc, argv);
   if (ret != NO_ERROR) return ret;
	
   if (DEBUG_LVL == 1) d1_printf3("%s %s ", ite_basename(inputfile), comment);

   return NO_ERROR;
}

int
ite_main_load(Tracer *&tracer)
{
   int ret = NO_ERROR;
	
   ret = ite_io_init();
   if (ret != NO_ERROR) return ret;
	
   /* read input file */
   ret = read_input(tracer);
   if (ret != NO_ERROR) return ret;

   return ret;
}

int
ite_main(Tracer *tracer)
{
   int ret = NO_ERROR;

	switch (formatout) {
    case 'n': break;
    case 'b': ret = solve(tracer); break;
    case 'w': ret = walkSolve(); break;
    case 'm': wvfSolve(); break;
    default: write_output(formatout, tracer);   
             ret = CONV_OUTPUT;
   }

   return ret;
}

int
ite_final(int ret, Tracer *tracer)
{

   if (s_expected_result[0]) {
      ret = check_expected_result(ret);
   }

   if (ite_counters[BDD_NODE_FIND]==0) ite_counters[BDD_NODE_FIND] = 1;
   d4_printf4("BDD Lookup Statistic: %ld/%ld (%f hit rate)\n", 
         (long)(ite_counters[BDD_NODE_FIND] - ite_counters[BDD_NODE_NEW]), 
         (long)(ite_counters[BDD_NODE_FIND]),
         1.0 * (ite_counters[BDD_NODE_FIND] - ite_counters[BDD_NODE_NEW]) / ite_counters[BDD_NODE_FIND]);
   d4_printf4("BDD Hash Table Lookup Statistic: %ld/%ld (%f steps)\n", 
         (long)(ite_counters[BDD_NODE_STEPS]),
         (long)(ite_counters[BDD_NODE_FIND]),
         1.0 * ite_counters[BDD_NODE_STEPS] / ite_counters[BDD_NODE_FIND]);
   d4_printf4("BDD Hash Table Storage Statistic: %ld/%ld (%f nodes taken)\n", 
         (long)(ite_counters[BDD_NODE_NEW]),
         (long)((1<<(numBuckets+sizeBuckets))-1),
         1.0 * ite_counters[BDD_NODE_NEW] / (long)((1<<(numBuckets+sizeBuckets))-1));

   char result_string[64];
   switch(ret) {
    case TRIV_SAT: 
       {
          strcpy(result_string, "Satisfiable");
          Verify_NoSolver(tracer);
       };
       break;
    case SOLV_SAT: {
       if (ite_counters[NUM_SOLUTIONS] > 1) {
          sprintf(result_string, "Satisfiable(%lld)", ite_counters[NUM_SOLUTIONS]);
       } else {
          strcpy(result_string, "Satisfiable");
       }
       Verify_Solver(tracer);
    } 
       break;
    case TRIV_UNSAT: 
    case SOLV_UNSAT: strcpy(result_string, "Unsatisfiable"); break;
    case SOLV_UNKNOWN: strcpy(result_string, "Unknown Result"); break;
    case CONV_OUTPUT: strcpy(result_string, "Conversion"); break;
    case SOLV_ERROR: strcpy(result_string, "SOLVER ERROR: Result is not as expected"); break;
    default: sprintf(result_string, "Error(%d)", ret); break;
   }

   ite_io_free();
   ite_main_free(tracer);

   ite_counters_f[RUNNING_TIME] = get_runtime();
   if (DEBUG_LVL == 1) {
      d1_printf3("%s %lld ", result_string, ite_counters[NO_ERROR]);
      d1_printf4("%.3fs %.3fs %.3fs\n", 
            ite_counters_f[RUNNING_TIME], 
            ite_counters_f[PREPROC_TIME], 
            ite_counters_f[BRANCHER_TIME]);
   } else {
      if (DEBUG_LVL > 0) {
         d0_printf2("%s\n", result_string);
         d2_printf2("Total Time: %4.3f \n", get_runtime());
      } else {
         printf("%s\n", result_string);
      }
   }

   if (ret == SOLV_ERROR) return 1;
   return 0;
}

int 
ite_pre_init()
{
   int i;
   for (i=0;i<MAX_COUNTER;i++) ite_counters[i] = 0;
   for (i=0;i<MAX_COUNTER_F;i++) ite_counters_f[i] = 0;

   init_terminal_in();
   init_terminal_out();

   return NO_ERROR;
}

void 
ite_main_free(Tracer *tracer)
{
   d9_printf1("ite_main_free\n");
   if (tracer != NULL) {
      delete tracer;
      d9_printf2("Removing file: %s\n", tracer_tmp_filename);
      unlink (tracer_tmp_filename);
   }

   delete [] variablelist;
   variablelist = NULL;
   delete [] original_functions;
   original_functions = NULL;
	ite_free((void **)&var_score);
	
   bdd_bdd_free_pools();
   bdd_circuit_free();
}

void 
ite_io_free()
{
   d9_printf1("ite_io_free\n");
   if (finputfile && finputfile != stdin) fclose(finputfile);
   if (foutputfile && foutputfile != stdout && foutputfile != stderr) fclose(foutputfile);
   //if (fresultfile != stdin && fresultfile) fclose(fresultfile);
}

int 
ite_io_init()
{
   d9_printf1("ite_io_init\n");

   /* 
    * open the output file 
    */

   if (!strcmp(outputfile, "-")) foutputfile = stdout;
   else {
      struct stat buf;
      if (stat(outputfile, &buf) == 0) {
         dE_printf2("Error: File %s exists\n", outputfile);
         return ERR_IO_INIT;
      }
      foutputfile = fopen(outputfile, "w");
   }

   if (!foutputfile) { 
      dE_printf2("Can't open the output file: %s\n", outputfile);
      return ERR_IO_INIT;
   } else d9_printf2("Output file opened: %s\n", outputfile);

   /* 
    * open the result file
    */
   /*
    if (input_result_filename[0]) {
    fresultfile = fopen(input_result_filename, "r");

    if (!fresultfile) {
    dE_printf2("Can't open the result filename: %s\n", input_result_filename);
    return ERR_IO_INIT;
    } else
    d9_printf2("Result file opened: %s\n", input_result_filename);
    }
    */
   return NO_ERROR;
}

int 
ite_preprocessing()
{
   d9_printf1("ite_preprocessing\n");
   int ret=0, prep_ret=0;
   /*
    * Preprocessing - start
    */
   prep_ret = Init_Preprocessing();

   /*
    * read partial/full assignment file
    */ 
   /*
    extern varinfo * variablelist;
    extern char input_result_filename[128];
    int read_input_result(FILE *, Tracer *tracer, varinfo*);
    FILE *fresultfile=NULL;

    if (input_result_filename[0]) {
    if (read_input_result(fresultfile, tracer, variablelist)) goto ExitNormal;
    else {
    d9_printf1("Finished reading input result\n");
   //ret = MakeInfs(variablelist);
   //ret = DoInfs();
   }
   }
   else d9_printf1("Result input file not specified\n");
    */

   if (prep_ret == PREP_NO_CHANGE || prep_ret == PREP_CHANGED) {
      prep_ret = Preprocessor (variablelist);
   }

   ret = Finish_Preprocessing();

   if (prep_ret == TRIV_SAT || prep_ret == TRIV_UNSAT) {
      ret = prep_ret;
   }
   /*
    * Preprocessing - end
    */
   d9_printf2("ite_preprocessing done: %d\n", ret);
   return ret;
}

int 
check_expected_result(int result)
{
   switch (result) {
    case TRIV_SAT: 
       if (strcmp(s_expected_result, "TRIV_SAT") && 
             strcmp(s_expected_result, "SAT")) return SOLV_ERROR;
       break;
    case TRIV_UNSAT: 
       if (strcmp(s_expected_result, "TRIV_UNSAT") && 
             strcmp(s_expected_result, "UNSAT")) return SOLV_ERROR;
       break;
    case SOLV_SAT:  
       if (strcmp(s_expected_result, "SOLV_SAT") && 
             strcmp(s_expected_result, "SAT")) return SOLV_ERROR;
       break;
    case SOLV_UNSAT: 
       if (strcmp(s_expected_result, "SOLV_UNSAT") && 
             strcmp(s_expected_result, "UNSAT")) return SOLV_ERROR;
       break;
    default: /* can't verify the result */
       dE_printf1("Can't check the result against the expected result\n");
       break;
   }	
   return result;
};
